### DISCUSSION TITLE
[Feature] Grey feature

### DISCUSSION THREAD

**Discussion by crazychengmm** [apache/skywalking#12085] [Q&A] [answered]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

I have an idea to integrate canary release functionality by incorporating tracing tags labeled sw8 into the configuration. Once certain rules are hit, these labels will be attached to the sw8 tracing headers. Istio can then leverage these tags to perform corresponding actions within the context of the canary rollouts, utilizing the full spectrum of tracing tags for precise traffic management.

### Use case

Is this idea mature? If possible, I would like to proceed with it.

### Are you willing to submit a pull request to implement this on your own?

- [X] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment]
I am a little confused about this asking. Are you asking for something we had verified before?
Check this, https://github.com/SphereEx/cyborg-flow

**Comment by crazychengmm** [comment]
Not same with [cyborg-flow](https://github.com/SphereEx/cyborg-flow).
<img width="756" alt="image" src="https://github.com/apache/skywalking/assets/23517542/63dd95ff-18ff-4694-84c0-dd78782f421e">

I want use sw8, work with istio. 

<img width="1242" alt="image" src="https://github.com/apache/skywalking/assets/23517542/523ec76b-68eb-4ef3-bcf8-195d206d8b24">

**Reply by wu-sheng** [reply]
If you have Istio, it has version tag to support route, why do you need an APM to do so?

**Reply by crazychengmm** [reply]
> If you have Istio, it has version tag to support route, why do you need an APM to do so?

Our problem is backend app do not tag the version on request, and we do not  want to put the logic on hard code.I suggest use skywalking agent to do the tag when first request with front page.

<img width="1081" alt="image" src="https://github.com/apache/skywalking/assets/23517542/42d8108c-f55e-4d39-96b2-0a43e0f17126">

**Reply by wu-sheng** [reply]
Collaboration context support auto tags. That is supported already.

- https://skywalking.apache.org/docs/skywalking-java/latest/en/setup/service-agent/java-agent/application-toolkit-trace-correlation-context/
- Use correlation.auto_tag_keys mentioned here, https://skywalking.apache.org/docs/skywalking-java/latest/en/setup/service-agent/java-agent/configurations/

**Comment by crazychengmm** [comment]
> Collaboration context support auto tags. That is supported already.
> 
> * https://skywalking.apache.org/docs/skywalking-java/latest/en/setup/service-agent/java-agent/application-toolkit-trace-correlation-context/
> * Use correlation.auto_tag_keys mentioned here, https://skywalking.apache.org/docs/skywalking-java/latest/en/setup/service-agent/java-agent/configurations/

Yes, That is what i am saying.

Here is my opinion.
custom context condition: 
    if userId = tom in http request head then
        TraceContext.putCorrelation("greyflow", "userId:tom");
If Skywalking have this config feature what custom context set with condition config, would be great.

**Reply by wu-sheng** [reply, marked as answer]
> if userId = tom in http request head 

This would not work in that way, but if you follow [SkyWalking Cross Process Correlation Headers Protocol](https://skywalking.apache.org/docs/main/latest/en/api/x-process-correlation-headers-v1/) and use `correlation.auto_tag_keys`, it works.
If you need to get the value of the header on Envoy, you may need a WASM plugin to support custom routines. And note `SkyWalking Cross Process Correlation Headers Protocol` is not supported by all agents, I just know Java and Golang agents support that. You may need to check other agents' docs if you need.

**Reply by crazychengmm** [reply]
Thank you. Help me a lot. Good day.
